File: app/src/main/java/com/karol/sezonnazdrowie/view/controls/TimePreference.java
package com.karol.sezonnazdrowie.view.controls;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;

import androidx.preference.DialogPreference;

public class TimePreference extends DialogPreference {

    public int hour;
    public int minute;

    public static int getHour(String time) {
        String[] pieces=time.split(":");

        return(Integer.parseInt(pieces[0]));
    }

    public static int getMinute(String time) {
        String[] pieces=time.split(":");

        return(Integer.parseInt(pieces[1]));
    }

    public TimePreference(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    @Override
    protected Object onGetDefaultValue(TypedArray a, int index) {
        return a.getString(index);
    }

    @Override
    protected void onSetInitialValue(boolean restoreValue, Object defaultValue) {
        String time;

        if (restoreValue) {
            if (defaultValue == null) {
                time = getPersistedString("20:00");
            } else {
                time = getPersistedString(defaultValue.toString());
            }
        } else {
            time = defaultValue.toString();
        }

        hour = getHour(time);
        minute = getMinute(time);
    }

    public static String timeToString(int h, int m) {
        return String.format("%02d:%02d", h, m);
    }

    public void persistStringValue(String value) {
        persistString(value);
    }
}

File: app/src/main/java/com/karol/sezonnazdrowie/view/controls/TimePreference.kt
package com.karol.sezonnazdrowie.view.controls

import android.content.Context
import android.content.res.TypedArray
import android.util.AttributeSet

import androidx.preference.DialogPreference
import org.threeten.bp.LocalTime

class TimePreference @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0,
    defStyleRes: Int = 0
) : DialogPreference(context, attrs, defStyleAttr, defStyleRes) {

    var time: LocalTime = LocalTime.of(20, 0)

    override fun onGetDefaultValue(a: TypedArray, index: Int): Any? = a.getString(index)

    override fun onSetInitialValue(restoreValue: Boolean, defaultValue: Any?) {
        val defaultAsLong = (defaultValue as String?)?.toLongOrNull() ?: 72000000000000L
        val nanoOfDay = if (restoreValue) {
            getPersistedLong(defaultAsLong)
        } else {
            defaultAsLong
        }

        time = LocalTime.ofNanoOfDay(nanoOfDay)
    }

    fun persistValue(value: LocalTime) {
        persistLong(value.toNanoOfDay())
    }
}
